README
Math-Polynomial version 1.012
=============================

Math::Polynomial is a Perl class representing polynomials in one
variable.  It provides a set of operations defined for polynomials,
like addition, multiplication, division with remainder, evaluation,
nesting, etc., as well as attribute inspection and formatting
capabilities.

More documentation is in the module file, available through perldoc.
Some additional usage examples can be found in the examples subdirectory.

DLSIP STATUS

   Development State:  R - Released
   Language Used:      p - Perl-only
   Support Level:      d - Developer
   Interface Style:    O - Object oriented
   Public License:     p - Standard-Perl

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

PROVIDED MODULES AND SCRIPTS

The distribution provides these modules:

  - Math::Polynomial          - main module
  - Math::Polynomial::Generic - interface extension (deprecated)

There are no scripts included as a part of the user interface.

DEPENDENCIES

This module requires these other modules and libraries to run:

  - perl version 5.6.0 or higher
  - overload        (usually bundled with perl)
  - Carp            (usually bundled with perl)

Additional requirements to run the test suite:

  - Test            (usually bundled with perl)

Recommended modules for increased functionality:

  - Math::Complex   (usually bundled with perl)
  - Math::BigNum    (available on CPAN)
  - Math::ModInt    (available on CPAN)

AUTHOR

Written and maintained by Martin Becker "becker-cpan-mp@cozap.com".

Inspired by the module of the same name written by Mats Kindahl
"mats@kindahl.net" 1997-2001, maintained by Martin Becker since 2007.

COPYRIGHT AND LICENCE

Copyright (c) 2007-2016 by Martin Becker.  All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Math::Polynomial.

                    Math::Polynomial Version 1.012
                    ------------------------------

- Release date: 2017-06-11

- Compatibility tests with alien modules do not trigger failures any more.
  Closes: [rt.cpan.org #114004], [rt.cpan.org #118895]

- Math::BigNum replaces Math::BigRat in recommendations and examples.

                    Math::Polynomial Version 1.011
                    ------------------------------

- Release date: 2016-08-08

- Metadata now recommends Math::BigRat 0.260804.

- Improved test diagnostics addressing a Math::BigRat issue
  [rt.cpan.org #114004].

                    Math::Polynomial Version 1.010
                    ------------------------------

- Not released.

                    Math::Polynomial Version 1.009
                    ------------------------------

- Release date: 2016-07-28

- Improved metadata.

- Added deprecation warning to Math::Polynomial::Generic.

                    Math::Polynomial Version 1.008
                    ------------------------------

- Release date: 2015-04-18

- Added wrap parameter for string_configuration.

                    Math::Polynomial Version 1.007
                    ------------------------------

- Release date: 2015-03-13

- Added exp_mod() method.

- Scheduled Math::Polynomial::Generic for deprecation.

                    Math::Polynomial Version 1.006
                    ------------------------------

- Release date: 2013-05-28

- Small documentation fixes.

                    Math::Polynomial Version 1.005
                    ------------------------------

- Release date: 2012-09-23

- Lifted a limitation on coefficients for differentiate() and integrate().

- Added an optimization for pow() and pow_mod().

                    Math::Polynomial Version 1.004
                    ------------------------------

- Release date: 2010-09-26

- Eliminated a perl-5.10-ism that had slipped into the test suite.

                    Math::Polynomial Version 1.003
                    ------------------------------

- Release date: 2010-09-08

- Added optimization for monic denominators to mmod.

- Added small optimization to nest.

- Deprecated optional parameter of div_root.

- Added as_horner_tree and as_power_sum_tree methods, discussed in
  [rt.cpan.org #46575].

- Added tests covering tree conversion methods.

- Added sign_of_coeff parameter for stringification and tree conversions.

- Simplified Horner example script.

- Added example script demonstrating Math::Symbolic interoperability.

- Added test covering Math::ModInt interoperability.

                    Math::Polynomial Version 1.002
                    ------------------------------

- Release date: 2009-06-11

- Improved interpolation polynomial calculation by one order of
  magnitude (O(n**3) => O(n**2)).

- Added optimization for monic denominators to div / mod / divmod.

- Added from_roots constructor and mul_root / div_root / divmod_root
  methods suggested by Kevin Ryde [rt.cpan.org #46427].

                    Math::Polynomial Version 1.001
                    ------------------------------

- Release date: 2009-05-19

- Repackaged after Math-Polynomial-1.000.tar.gz had file transmission
  problems on its way to CPAN.

                    Math::Polynomial Version 1.000
                    ------------------------------

- Release date: 2009-05-19

- Author: Martin Becker

- Rewritten from scratch, using
  h2xs 1.23 with options -X -A -b 5.6.0 Math::Polynomial

- Usable coefficient spaces are now any object classes with
  well-defined basic arithmetic operations and equality checks.

- Objects are now essentially immutable.  The only property that
  may be changed dynamically configures formatting defaults.
  Thus, cloning is hardly ever necessary.

- Removed tidy() and size() methods and dropped tidiness concept, as it
  bore no mathematical significance.

- Degrees of coefficients in new() are now in ascending order and
  therefore equivalent to array indexes.

- Added monomial constructor.

- Changed Lagrange interpolation to take arrays of x- and y-values.

- Extended constructors to borrow coefficient space and stringification
  attributes when used as object methods.

- Added usage of coeff() without parameter, yielding list of all coefficients.

- Added coefficients() method, yielding non-empty list of coefficients.

- Added coeff_zero() and coeff_one() methods, yielding zero and unit
  elements of coefficient space.

- Added proper_degree() method yielding undef for zero polynomials.

- Added is_zero() method yielding true for zero polynomials.

- Renamed quotrem() operator to divmod().

- Added overloaded operators: bool ! ** << >> == !=

- Added methods: is_nonzero(), is_equal(), is_unequal(), is_monic(),
  add_const(), sub_const(), mul_const(), div_const(), pow(), pow_mod(),
  inv_mod(), shift_up(), shift_down(), slice(), monize(), mmod(), gcd(),
  xgcd(), differentiate(), integrate(), definite_integral().

- Renamed eval() method to evaluate() and sub() method to sub_().

- Removed configure() class method, renamed to_string() method to
  as_string() and added optional configuration parameter.

- Added lots of configuration options for string representation.

- Added string_config() method to configure as_string default behaviour
  globally and per object.

- Added $max_degree class variable to configure parameter sanity tests.

- Changed internal object structure to improve extensibility.

- Reworked test suite and examples.

- Reworked pod documentation.

- Added experimental module Math::Polynomial::Generic.


                    Math::Polynomial Version 0.04
                    -----------------------------

- Release date: 2007-07-12

- New maintainer: Martin Becker.

- Bugfix [rt.cpan.org #4788] (reported 2003 by prospective maintainer):
  Method tidy() looped endlessly on zero polynomials.

- Bugfix: Crashed on adding a constant.


t/lib/Test/MyUtils.pm
# Copyright (c) 2008-2017 Martin Becker.  All rights reserved.
# This package is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

# Utility functions for tests:
# * conditionally skip tests if required modules are not available
# * conditionally skip tests intended for the maintainer only
# * conditionally skip tests if some file cannot be read into a string
# * fetch an executable file name of the perl binary currently running

package Test::MyUtils;

use 5.006;
use strict;
use warnings;
use Config;
use base 'Exporter';

our $VERSION   = '0.008';
our @EXPORT    = qw(use_or_bail maintainer_only);
our @EXPORT_OK = qw(
    slurp_or_bail this_perl report_version
    init_comp_check comp_ok
);
our %EXPORT_TAGS = (
    default => \@EXPORT,
    comp    => [qw(init_comp_check comp_ok)],
);

our $DIST_NAME    = _guess_distname();
our $MAX_FILESIZE = 1024 * 1024;

my $comp_failure  = 'no good reason';

sub _guess_distname {
    my $distname = undef;
    if (open my $rh, '<', 'README') {
        my $headline = <$rh>;
        if (defined $headline && $headline =~ /^\s*(\w+(?:[^\w\s]+\w+)*)\s/) {
            $distname = $1;
        }
        close $rh;
    }
    if (!defined $distname) {
        $distname = 'This-Distribution';
    }
    return $distname;
}

sub _skip_all {
    my ($reason) = @_;
    print "1..0 # SKIP $reason\n";
    exit 0;
}

# To enforce a minimum version of a module, supply a $version value.
# To use a module with default imports, omit $imports_ref.
# To use a module with explicit or no imports, supply an array reference.
sub use_or_bail {
    my ($module, $version, $imports_ref) = @_;

    if (!eval "require $module") {
        _skip_all("$module not available");
    }

    my $loaded_version = eval { $module->VERSION };
    if (defined $loaded_version) {
        print "# module $module version is $loaded_version\n";
    }

    if (defined($version) && !defined eval { $module->VERSION($version) }) {
        _skip_all("$module version $version or higher not available");
    }

    if (!$imports_ref || @{$imports_ref}) {
        my $package = caller;
        my @imports = $imports_ref? @{$imports_ref}: ();
        if (!eval "package $package; \$module->import(\@imports); 1") {
            my $error = $@;
            $error =~ s/\n.*//s;
            _skip_all("import of $module failed: $error");
        }
    }
    return 1;
}

# Call this before plan() in test scripts reserved for the maintainer.
# Add names of mandatory configuration features for further restrictions.
sub maintainer_only {
    my @required_features = @_;
    my $env_maint = 'MAINTAINER_OF_' . uc $DIST_NAME;
    $env_maint =~ s/[_\W]+/_/g;
    if (!$ENV{$env_maint}) {
        _skip_all("setenv $env_maint=1 to run these tests");
    }
    foreach my $feature (@required_features) {
        if (!$Config{$feature}) {
            _skip_all("feature not available: $feature");
        }
    }
    return 1;
}

# Call this before plan() in test scripts analysing some file.
# Return value is the file content.  Returns only on success.
sub slurp_or_bail {
    my ($filename) = @_;
    local $/;
    my $fh;
    my $result;
    my $err;
    if (!-e $filename) {
        $err = 'file does not exist';
    }
    elsif (!-f _) {
        $err = 'not a plain file';
    }
    elsif ($MAX_FILESIZE < -s _) {
        $err = 'file too large';
    }
    elsif (open $fh, '<', $filename) {
        defined($result = <$fh>) or $err = "cannot read: $!";
        close $fh;
    }
    else {
        $err = "cannot open: $!";
    }
    if (!defined $result) {
        _skip_all("$filename: $err");
    }
    return $result;
}

sub this_perl {
    my $this_perl = $Config{'perlpath'};
    my $suffix    = exists($Config{'_exe'})? $Config{'_exe'}: '';
    if ($^O ne 'VMS' && '' ne $suffix && $this_perl !~ /$suffix\z/) {
        $this_perl .= $suffix;
    }
    return $this_perl;
}

sub report_version {
    foreach my $module (@_) {
        my $text = 'can not be loaded';
        if (eval "require $module") {
            my $version = eval { $module->VERSION };
            if (defined $version) {
                $text = "version is $version";
            }
            else {
                $text = 'has no version';
            }
        }
        print "# module $module $text\n";
    }
}

sub init_comp_check {
    my ($ours, @alien) =
        map {
            my $v = $_->VERSION;
            "$_ v$v"
        } @_;
    my $theirs    = join q[ + ], @alien;
    my $is        = 1 == @alien? 'is': 'are';
    $comp_failure = "$theirs $is not compatible with $ours";
    die qq{use "Test" before initializing compatibility tests\n}
        if !exists $INC{'Test.pm'};
}

sub comp_ok {
    my ($ok, $topic) = @_;
    print "# failure with $topic\n" if !$ok && defined $topic;
    Test::skip(!$ok && $comp_failure, $ok);
}

1;
__END__


examples/horner.pl
#!/usr/bin/perl

# Copyright (c) 2009 Martin Becker.  All rights reserved.
# This package is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
#
# $Id: horner.pl 4 2009-05-10 22:14:58Z demetri $

# Math::Polynomial example: write polynomials in the form of a Horner scheme
#
# A Horner scheme is an efficient method to evaluate a polynomial for
# a value of x = x0 by repeatedly adding a coefficient and multiplying
# by x0.  This script defines a function horner() turning a polynomial
# into an equivalent Horner-like expression.  It displays the output
# of this function for a couple of simple polynomials.

use strict;
use warnings;
use Math::Polynomial 1.000;

foreach my $a (-1, 0, 1, 2) {
    foreach my $b (-1, 0, 1) {
        foreach my $c (0, 1) {
            foreach my $d (0, 1) {
                my $p = Math::Polynomial->new($d, $c, $b, $a);
                print horner($p), "\n";
            }
        }
    }
}

sub horner {
    my ($p) = @_;
    my $result = undef;
    my $sign = 0;
    my $zero = $p->coeff_zero;
    my $one  = $p->coeff_one;
    foreach my $c (reverse $p->coefficients) {
        if (defined $result) {
            if ($sign) {
                $result = "($result)";
            }
            $result .= '*x';
            $sign = $c <=> $zero;
            if ($sign < 0) {
                $result .= $c;
            }
            elsif ($sign > 0) {
                $result .= "+$c";
            }
        }
        else {
            $result = "$c";
        }
    }
    $result =~ s/((?:^|\()-?)1\*/$1/;       # optimize: 1*x => x
    return $result;
}

__END__


examples/migration.pl
#!/usr/bin/perl

# Copyright (c) 2009 Martin Becker.  All rights reserved.
# This package is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
#
# $Id: migration.pl 12 2009-05-14 07:34:36Z demetri $

# Math::Polynomial usage example: Migration from version 0.04 to 1.000
#
# Math::Polynomial version 1.000 broke backwards compatibility with
# earlier versions.  This example demonstrates how legacy code using
# Math::Polynomial version 0.04 can be adapted to use a more current
# version of Math::Polynomial.
#
# Throughout the rest of this example, old code is in comments,
# immediately followed by changed code.  Parts needing no modification
# are in blocks without any comments.

use strict;
use warnings;

# use Math::Polynomial;
use Math::Polynomial 1.000;

# my $p = Math::Polynomial->new(1, 3, -2);
my $p = Math::Polynomial->new(-2, 3, 1);

# Math::Polynomial->verbose(0);
Math::Polynomial->string_config({
    with_variable => 0,
    plus          => q{ },
});

print "p is $p\n";

# my $result = $p->eval(5);
my $result = $p->evaluate(5);

print "p(5) is $result\n";

# my $q = Math::Polynomial->new(2, 3);
my $q = Math::Polynomial->new(3, 2);

# my $r = $q->clone;
my $r = $q;

$r += 1;
print "q is $q, r is $r\n";

# my $s = $q - $r;
# $s->tidy;
my $s = $q - $r;

print "s is $s\n";

# Math::Polynomial->verbose(1);
Math::Polynomial->string_config({
    fold_sign     => 1,
    leading_minus => q{-},
    times         => q{*},
    variable      => q{$X},
    power         => q{**},
    prefix        => q{},
    suffix        => q{},
});

# my ($quot, $rem) = Math::Polynomial::quotrem($p, $q);
my ($quot, $rem) = $p->divmod($q);

print "($p) / ($q) = $quot\n";
print "($p) % ($q) = $rem\n";

# print "q^2 % p = ", $q * $q % $p, "\n";
print "q^2 % p = ", $q**2 % $p, "\n";

# my $pp = Math::Polynomial::interpolate(1 => 9, 2 => 14, 3 => 11);
my $pp = Math::Polynomial->interpolate([1, 2, 3], [9, 14, 11]);

print "pp = $pp\n";

__END__


lib/Math/Polynomial/Generic.pm
# Copyright (c) 2009-2017 by Martin Becker.  All rights reserved.
# This package is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

package Math::Polynomial::Generic;

use 5.006;
use strict;
use warnings;
use Carp qw(carp croak);
use Math::Polynomial 1.000;

require Exporter;

use base qw(Math::Polynomial Exporter);

our @EXPORT_OK = qw( X C );
our %EXPORT_TAGS = ( legacy => [] );

our $VERSION = '1.012';

my $deprecated  = 1;
my $x_singleton = Math::Polynomial::Generic->new(0, 1);

sub X () {
    _warn_deprecated() if $deprecated;
    return $x_singleton;
}

sub C ($) {
    my ($value) = @_;
    return Math::Polynomial->new($value);
}

sub import {
    $deprecated = 1;
    foreach my $arg (@_[1 .. $#_]) {
        $deprecated = 0, last if ':legacy' eq $arg;
    }
    _warn_deprecated() if $deprecated;
    Math::Polynomial::Generic->export_to_level(1, @_);
}

sub _warn_deprecated {
    $deprecated = 0;
    carp __PACKAGE__, ' is deprecated';
}

sub _is_generic { return 1; }

sub _cast {
    my ($this, $that) = @_;
    my ($zero, $one) = ($that->coeff_zero, $that->coeff_one);
    my @coeff =
        map {
            my $n = $_;
            my ($r, $c) = ($zero, $one);
            if ($n < 0) {
                $n = -$n;
                $c = -$c;
            }
            while ($n) {
                if (1 & $n) {
                    $r = $r + $c;
                }
                $n >>= 1 and $c = $c + $c;
                # note: "+=" intentionally avoided
            }
            $r
        } $this->coeff;
    return $that->new( @coeff );
}

sub divmod {
    croak 'implementation restriction: generic division not defined';
}

sub div { goto &divmod; }
sub mod { goto &divmod; }

1;
__END__

=head1 NAME

Math::Polynomial::Generic - syntactical sugar coating Math::Polynomial

=head1 VERSION

This documentation refers to version 1.012 of Math::Polynomial::Generic.

=head1 DEPRECATION NOTICE

This module has always been declared as experimental and its interface
as not to be taken for granted.  The intention was to give users
more expressive flexibility through a kind of polynomial object
that is not bound to a particular coefficient space.

This experiment is now considered a failure.  Coefficient spaces need to
be specified at one point, and delaying this part of proper initialization
seems do do more harm than good in terms of clarity.

Therefore, Math::Polynomial::Generic is now deprecated.  Using the module
now triggers a deprecation warning unless the special symbol C<:legacy>
appears on the import list.  At some time after July 1, 2017, the
module will be discontinued altogether.

=head1 SYNOPSIS

  use Math::Polynomial::Generic qw(:legacy X C);

  $p = X**2 - 3 * X + 5;
  $q = $p * X;
  $r = (X - $some_value) * (X - $some_other_value);
  $s = C($some_value) * X**2 + C($some_other_value);

=head1 DESCRIPTION

Math::Polynomial::Generic allows to create Math::Polynomial objects
in a more descriptive way than with basic constructors.  It offers
a symbol I<X> that will act as a polynomial when used as the variable
in a polynomial expression.  Another one-letter symbol I<C> turns
constants into constant polynomials.

=head2 SUBROUTINES

=over 4

=item I<X>

C<X> is different from C<Math::Polynomial-E<gt>new(0, 1)> in that
it is not bound to a particular coefficient space.  C<X> can be
coupled in expressions with polynomials of arbitrary coefficient
types, such as complex numbers, big rationals, square matrices,
etc.  The coefficients actually used determine the coefficient space
of the whole expression.  Incompatible coefficient types must not
be mixed in a single expression, of course.

The mechanism to make this work is based on a kind of generic object
(hence the name) which will be cast to a proper polynomial when it
is used in a binary operation together with something already bound
to a coefficient space: either another polynomial or a plain
coefficient.

=item I<C>

C<C($coeff)> creates a constant polynomial from a given coefficient
value C<$coeff>.

Coefficients other than simple numerical values should be turned
into polynomials to prevent perl from carrying out the overloaded
operator in the coefficient class rather than the polynomial class
(see below).

Expressions containing I<X> but lacking any coefficient values will
produce generic polynomial objects.  These must not be mistaken for
proper polynomials, nor should Math::Polynomial methods be invoked
on them.

In order to turn an otherwise generic expression into a regular
polynomial object, add a C<C()>-wrapped zero value.

=back

=head2 EXAMPLES

  $c = Math::BigNum->new('2/3');   # some coefficient value
  $p = Math::Polynomial->new(1);   # some regular polynomial

  $q = X;                          # wrong (generic)
  $q = X * X - X;                  # wrong (generic)
  $q = X * X - X + C(0);           # OK
  $q = $p + X;                     # OK
  $q = X + $p;                     # OK
  $q = $p - X;                     # OK
  $q = X - $p;                     # OK
  $q = $p * X;                     # OK
  $q = $p / X;                     # OK
  $q = $p % X;                     # OK

  $q = $c + X;                     # wrong (operand types)
  $q = X + $c;                     # wrong (operand types)
  $q = C($c) + X;                  # OK
  $q = X + C($c);                  # OK
  $q = C($c) * X**2;               # OK

  $q = X / X + C(0);               # wrong (generic division)

=head2 OVERRIDDEN METHODS

=over 4

=item I<divmod>

=item I<div>

=item I<mod>

Currently, division of generic objects by generic objects (like C<X/X>) is
not implemented.  The methods I<divmod>, I<div> and I<mod> are overridden
to guard against such cases.

=back

=head2 PROTECTED METHODS

=over 4

=item I<_is_generic>

Boolean true for generic objects, false for regular polynomials.

=item I<_cast>

C<$p-E<gt>_cast($q)> generates a regular polynomial from a generic
object C<$p>, the result sharing the coefficient space with C<$q>.

=back

=head2 EXPORT

By default, nothing is exported into the caller's namespace.  The
polynomial generators X and C can be explicitly imported, however.

The special symbol C<:legacy> can be imported in order to suppress
deprecation warnings.

=head2 DIAGNOSTICS

=over 4

=item Math::Polynomial::Generic is deprecated

This warning is issued when the module is imported without the C<:legacy>
tag in the import list or when Math::Polynomial::Generic::X is called
for the first time after the module has been loaded without an import.

=back

=head2 MIGRATION

As this module has been declared deprecated, existing applications
should be migrated to no longer use it.  This, fortunately, is
rather easy.  Instead of the symbol C<X>, we recommend to use a
variable C<$X> that is initialized as
C<Math::Polynomial-E<gt>new($zero, $one)> with appropriate coefficient
values C<$zero> and C<$one>.

C<C($coeff)> can be replaced by C<Math::Polynomial-E<gt>new($coeff)>
or it can be defined locally as a small wrapper for the same.

=head1 SEE ALSO

  Math::Polynomial

=head1 AUTHOR

Martin Becker, E<lt>becker-cpan-mp@cozap.comE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2009-2017 by Martin Becker.  All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.

=head1 DISCLAIMER OF WARRANTY

This module is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut


Makefile.PL
# Copyright (c) 2008-2017 Martin Becker.  All rights reserved.
# This package is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

use strict;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME           => 'Math::Polynomial',
    VERSION_FROM   => 'lib/Math/Polynomial.pm',
    PREREQ_PM      => {
        overload   => 0,
        Carp       => 0,
    },
    ABSTRACT_FROM  => 'lib/Math/Polynomial.pm',
    AUTHOR         => 'Martin Becker <becker-cpan-mp@cozap.com>',
    $ExtUtils::MakeMaker::VERSION >= 6.18? (
        SIGN       => 1,
    ): (),
    $ExtUtils::MakeMaker::VERSION >= 6.31? (
        LICENSE    => 'perl',
    ): (),
    $ExtUtils::MakeMaker::VERSION >= 6.46? (
        META_MERGE => {
            'meta-spec' => { version => 2 },
            prereqs => {
                runtime => {
                    recommends => {
                        'Math::Complex'       => 0,
                        'Math::BigNum'        => 0,
                        'Math::ModInt'        => 0,
                    },
                },
            },
            provides => {
                'Math::Polynomial' => {
                    file    => 'lib/Math/Polynomial.pm',
                    version => '1.012',
                },
                'Math::Polynomial::Generic' => {
                    file    => 'lib/Math/Polynomial/Generic.pm',
                    version => '1.012',
                },
            },
        },
    ): (),
    $ExtUtils::MakeMaker::VERSION >= 6.47? (
        MIN_PERL_VERSION => '5.006',
    ): (),
    $ExtUtils::MakeMaker::VERSION >= 6.56? (
        BUILD_REQUIRES => {
            'Test'                => 0,
            'ExtUtils::MakeMaker' => 0,
        },
    ): (),
);
